colossalai/zero/low_level/low_level_optim.py

```diff
@@ -193,486 +109,76 @@ def __init__(
-
-        for param in param_list:
-            padding_size = (
-                self._bucket_store.zero_world_size - param.numel() % self._bucket_store.zero_world_size
-            ) % self._bucket_store.zero_world_size
-            self._param_store.record_param_padding_size(param, padding_size)
-
-            with torch.no_grad():
-                if padding_size > 0:
-                    padding_param = torch.nn.functional.pad(param.data.view(-1), [0, padding_size])
-                    # reset working params' ptr when no master weights
-                    if self._master_weights == False:
-                        param.data = padding_param[: param.numel()].view(param.shape)
-                else:
-                    padding_param = param.data.view(-1)
-
-                if self._bucket_store.moe_extra_dp_pg is not None and is_moe_tensor(param):
-                    splited_params = padding_param.split(
-                        padding_param.numel() // self._bucket_store.moe_extra_dp_pg_size
-                    )
-                    splited_params = splited_params[self._bucket_store.moe_extra_dp_pg_rank]
-                else:
-                    splited_params = padding_param.split(padding_param.numel() // self._bucket_store.zero_world_size)
-                    splited_params = splited_params[self._bucket_store.zero_local_rank]
-
-                # use fp32 when master_weights is True
-                if self._master_weights is True:
-                    splited_param_current_rank = splited_params.detach().float().to(device)
-                else:
-                    splited_param_current_rank = splited_params
-
-                # Send the splited view to the optimizer to match ZeRO 2 grad shape
-                params_current_rank.append(splited_param_current_rank)
-                self._param_store.link_master_and_working_param(splited_param_current_rank, param)
-
-        return params_current_rank
-
-    ###########################
-    # Backward Reduction Hook #
-    ###########################
-
-    @staticmethod
-    def grad_handler(
-        param: nn.Parameter,
-        group_id: int,
-        bucket_store: BucketStore,
-        param_store: ParameterStore,
-        grad_store: GradientStore,
-    ):
-        # if run with no_sync context, would not sync grad when backward
-        if grad_store.require_grad_sync:
-            LowLevelZeroOptimizer.add_to_bucket(param, group_id, bucket_store, param_store, grad_store)
-
-    def _attach_reduction_hook(self):
-        # we iterate over the working params
-        # on each param, we register a hook to its AccumulateGrad object
-        for group_id in range(self.num_param_groups):
-            param_group = self._working_param_groups[group_id]
-            for param in param_group:
-                if param.requires_grad:
-                    param._grad_handle = param.register_post_accumulate_grad_hook(
```

Comment:
`register_post_accumulate_grad_hook` is essential for vae training.

Reply:
fix in #5824